### DISCUSSION TITLE
No explanation for `app/Models` directory in v8?

### DISCUSSION THREAD

**Discussion by SevenOutman** [laravel/framework#40142] [General] [answered]
Hi, I'm recently upgrading my Laravel app from v7 to v8, and I noticed that the scaffolding [now contains the `app/Models` directory](https://laravel.com/docs/8.x/structure#the-models-directory) which was [claimed to be missing intentionally in versions before v8](https://laravel.com/docs/7.x/structure#introduction). There's nothing mentioning this change in v8 upgrade guide so I don't know if the officially recommended approach has changed in v8 and whether I should move my model classes into this `app/Models` directory.

What is worse is that that when I moved model class files into `app/Models` directory and change their namespaces to `App\Models`, it breaks existing morph relations immediately because the full class names were persisted in database. By the time I notice this problem, there're already new morph relation records using the new `App\Models` class names in my database. I have no way to either continue or revert this change without manually update my existing data records. And again there's no documentation mentioning this.

**Comment by devcircus** [comment, marked as answer]
If I'm not mistaken, there's no reason to change anything. I think the "make:model" command will check to see if you have an "app/models" directory, and if so, it will place models there, otherwise it will place them in the "app" directory as in previous versions. 

I assume there was no reason to mention in the upgrade guide since it's not necessary to change anything.

**Reply by SevenOutman** [reply]
Thanks for your answer. But I believe it's still worth mentioning what users *should* do when they get confused why the "intentionally missing" directory is then added in the official scaffolding. For now I still need additional explanation outside of documentation like this answer from you to be certain that I don't have to consider `app/Models` as a best practice recommended by official.

**Reply by dennisprudlo** [reply]
I would say app/Models is a best practice which is recommended/documented as such: https://laravel.com/docs/8.x/structure#the-models-directory

**Reply by SevenOutman** [reply]
@dennisprudlo The key point in my view is that the v7 docs just said they don't use Models directory as an intentional decision. Then what factor makes them suddenly change their minds? If the factor is so important then should users follow this new pattern and update their directory structure? And how should users adopt this migration without breaking their existing app? Those should be explained in documentation, not let users guess IMO.

**Reply by crynobone** [reply]
Did you read https://laravel.com/docs/8.x/releases#models-directory

**Reply by dennisprudlo** [reply]
@SevenOutman I guess the reason the upgrade guide doesn't contain a section informing you about the new `app/Models` directory is, that it's simply not necessary to update the models. If you move alle your models from `app` to `app/Models` you will have to update all used model namespaces in other classes. And even if your IDE takes care of that sometimes you may use string concatenation to create a namespace path and your IDE may not identify it.

**Reply by SevenOutman** [reply]
@crynobone Thanks! I remember seeing this section when I was trying to upgrade from v6 (but ended up v7 at last). I searched "models" in upgrade guide but got no results, not realizing it was in the release notes. I have been always speaking highly of the clear and instructive upgrade guide of Laravel so I got so confused this time.

**Reply by SevenOutman** [reply]
@dennisprudlo That makes sense, thanks! I was just trying to keep everything in sync with the latest scaffolding when upgrading.
